dashboard/app/models/concerns/partial_registration.rb

```diff
@@ -7,62 +7,53 @@
   extend ActiveSupport::Concern
 
-  USER_ATTRIBUTES_SESSION_KEY = 'devise.user_attributes'
-  OAUTH_PARAMS_TO_STRIP = %w(oauth_token oauth_refresh_token)
+  SESSION_KEY = 'partial_registration'
 
   module ClassMethods
-    def new_from_partial_registration(session)
+    def new_from_partial_registration(session, &block)
       raise 'No partial registration was in progress' unless PartialRegistration.in_progress? session
-      new(session[USER_ATTRIBUTES_SESSION_KEY]) do |user|
-        cache = CDO.shared_cache
-        OAUTH_PARAMS_TO_STRIP.each do |param|
-          next if user.send(param)
-          # Grab the oauth token from memcached if it's there
-          cache_key = PartialRegistration.cache_key(param, user)
-          user.send("#{param}=", cache.read(cache_key)) if cache
-        end
-        yield user if block_given?
-      end
+      cache_key = session[SESSION_KEY]
+      json = CDO.shared_cache.read(cache_key)
+      attributes = JSON.parse(json)
+      new(attributes, &block)
     end
   end
 
   def self.in_progress?(session)
-    !!session[USER_ATTRIBUTES_SESSION_KEY]
+    session[SESSION_KEY] && CDO.shared_cache.exist?(session[SESSION_KEY])
   end
 
   def self.persist_attributes(session, user)
-    user = user.dup
-    # Because some oauth tokens are quite large, we strip them from the session
-    # variables and pass them through via the cache instead - they are pulled out again
-    # in new_from_partial_registration
-    cache = CDO.shared_cache
-    if cache
-      OAUTH_PARAMS_TO_STRIP.each do |param|
-        param_value = user.attributes['properties'].delete(param)
-        cache_key = PartialRegistration.cache_key(param, user)
-        cache.write(cache_key, param_value)
-      end
-    end
-    session[USER_ATTRIBUTES_SESSION_KEY] = user.attributes
-  end
-
-  def self.get_provider(session)
-    in_progress?(session) ? session[USER_ATTRIBUTES_SESSION_KEY]['provider'] : nil
+    # Push the potential user's attributes into our application cache.
+    cache_key = PartialRegistration.cache_key(user)
+    CDO.shared_cache.write(cache_key, user.attributes.compact.to_json)
+
+    # Put the cache key into the session, to
+    # 1. track that a partial registration is in progress
+    # 2. retrieve the attributes later when we're ready to persist the user
+    session[SESSION_KEY] = cache_key
   end
 
   def self.cancel(session)
     provider = get_provider(session) || 'email'
     SignUpTracking.log_cancel_finish_sign_up(session, provider)
     SignUpTracking.end_sign_up_tracking(session)
-    session.delete(USER_ATTRIBUTES_SESSION_KEY)
-    session
+    session.delete(SESSION_KEY)
+  end
+
+  def self.get_provider(session)
+    # Extract the provider name from the cache key to avoid actually
+    # interacting with the cache or doing deserialization.
+    # Assumption: Provider names will not contain hyphens
+    cache_key = session[SESSION_KEY]
+    /^([^-]+)-.+-partial-sso$/.match(cache_key)&.[](1)
```

Comment:
nit: I guess everyone should know this syntax, but `&.first` might read more easily

Reply:
oh -- this is different from `first` because you're getting the second element. ok, makes sense!

Reply:
Ruby offers a whole bunch of different ways to write this pattern:

1. `/^([^-]+)-.+-partial-sso$/.match(cache_key)&.[](1)`
2. `/^([^-]+)-.+-partial-sso$/.match(cache_key)&.captures&.first`
3. `/^([^-]+)-.+-partial-sso$/.match(cache_key) && $1`
4. `/^(?<match>[^-]+)-.+-partial-sso$/ =~ cache_key && match`
5. `/^([^-]+)-.+-partial-sso$/.match(cache_key)&.to_a&.second`
6. `/^([^-]+)-.+-partial-sso$/.match(cache_key)&.to_a&.last`
7. `/-.+-partial-sso$/.match(cache_key)&.pre_match`
8. `cache_key.scan(/^([^-]+)-.+-partial-sso$/).flatten.first`

None of them are all that great, but it's interesting to look at all the different options!

Reply:
:joy: Thanks Will!  I do like some of these better than others.  `.captures.first` seems especially explicit about its intent, which is nice.

If we're willing to trade one confusing syntax for another, I found even more options including using a lookahead to avoid captures altogether, and [this `str[regexp]` syntax](https://ruby-doc.org/core-2.5.0/String.html#method-i-5B-5D) that I've never seen before.

1. `/^[^-]+(?=-.+-partial-sso$)/.match(cache_key)&.to_s`
2. `cache_key[/^[^-]+(?=-.+-partial-sso$)/]`
3. `cache_key[/^([^-]+)-.+-partial-sso$/, 1]`

Reply:
I've updated the PR to use

```ruby
/^([^-]+)-.+-partial-sso$/.match(cache_key)&.captures&.first
```